Update to alpine and dockerize

Xaseron opened the pull request:
Benefits:
 - Smaller images 200MiB vs 700MiB
 - Faster build times (`mvn clean install -PbuildDocker -DskipTests `) 90s vs 120s
 - dockerize will fail if the timeout is exceeded

The problem with the wait-for-it.sh script is that it will not fail if the timeout is exceeded and contiune with the execution even if the requested service is not available. 

Exposing ports in the docker-compose file is only necessary if the port is accessed outside the docker network and that is only the case for the gateway service.

Xaseron committed:
Add alpine jre and dockerize script

Xaseron committed:
Add docker build argument DOCKERIZE_VERSION

Xaseron committed:
Increase memory limit

Xaseron committed:
Use dockerize instead of wait-for-it script

Xaseron committed:
Update documentation dockerize script

Xaseron committed:
Remove unnecessary ports

arey commented on docker-compose.yml:
```diff
@@ -3,14 +3,14 @@ services:
     image: mszarlinski/spring-petclinic-config-server
     container_name: config-server
-    mem_limit: 256M
+    mem_limit: 1024M
```
What is the reason to increase the memory limit from 256M to 1024M ?

Xaseron replied:
Yes, it works and i have changed the pull request accordingly.

arey reviewed (commented):
Thank you @Xaseron for your contribution. 
I'm aware to update to `alpine` and use `dockerize` (I didn't know this project) instead of  the `wait-for-it.sh` script.

For ports removal, it could be helpful for debugging to reach the REST API. And we have to access to server UI (ie. Zipkin, SPring Boot Admin, Server Config, Eureka)

I also have a question : why do you inscrease memory limit from 256M to 1024 ?

Xaseron committed:
Revert "Remove unnecessary ports"

This reverts commit a605aaf

Xaseron committed:
Set memory limit to 512M

arey commented:
@mszarlinski could you also review this PR please?

Xaseron commented:
The config container and simply crashes during startup with 256M memory limit

Xaseron commented:
% uname -a
```
Linux arch 4.18.12-arch1-1-ARCH #1 SMP PREEMPT Thu Oct 4 01:01:27 UTC 2018 x86_64 GNU/Linux
```
% docker version
```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.11
 Git commit:        e68fc7a215
 Built:             Fri Sep  7 11:26:59 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.11
  Git commit:       e68fc7a215
  Built:            Fri Sep  7 11:26:11 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

mszarlinski commented:
```
System Version: macOS 10.13.6 (17G65)
Kernel Version: Darwin 17.7.0
```

Xaseron commented:
Could you post a `docker stats` of the running microservices

mszarlinski commented:
Here is `docker stats` output for different OS images:

- openjdk:8 (mem_limit: 256M)
```
CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
de5e5e96108d        api-gateway         0.31%               255.2MiB / 256MiB   99.69%              36.3kB / 370kB      74.2MB / 78.6MB     45
24b32e76ef0c        visits-service      1.12%               148MiB / 256MiB     57.81%              65.7kB / 64.5kB     477MB / 277MB       46
9584d2076b76        customers-service   0.28%               206.1MiB / 256MiB   80.52%              66.3kB / 65.3kB     472MB / 303MB       46
bf9ab7c15741        vets-service        0.81%               172.4MiB / 256MiB   67.33%              67.2kB / 65.1kB     493MB / 266MB       46
23f8848a9195        tracing-server      1.01%               200.6MiB / 256MiB   78.36%              66.6kB / 67.5kB     38.4MB / 13MB       65
933f98f02051        admin-server        5.33%               221.4MiB / 256MiB   86.48%              136kB / 154kB       157MB / 57MB        50
ea26fdc66470        discovery-server    0.32%               202.8MiB / 256MiB   79.24%              203kB / 144kB       85.6MB / 85.2MB     42
4c92de729b7a        config-server       0.31%               161.3MiB / 256MiB   63.00%              149kB / 106kB       328MB / 28.3MB      36
```
- openjdk:8-jre-alpine (mem_limit: 512M)
```
CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
03c85a6963a0        api-gateway         1.06%               260.9MiB / 512MiB   50.95%              91.4kB / 401kB      69.7MB / 0B         66
57de68fbbfd8        visits-service      0.44%               289.6MiB / 512MiB   56.55%              14.8kB / 13.1kB     112MB / 0B          56
9e8d49f76ecb        customers-service   0.64%               302.2MiB / 512MiB   59.02%              16kB / 16kB         114MB / 0B          55
b55c5b57659d        vets-service        0.41%               295.9MiB / 512MiB   57.79%              15.5kB / 13.2kB     111MB / 0B          57
3698dd89a836        tracing-server      0.29%               131.8MiB / 512MiB   25.74%              16.8kB / 15.3kB     22.3MB / 0B         52
65f3bcab1080        admin-server        0.59%               197MiB / 512MiB     38.47%              48.6kB / 1.51MB     94.4MB / 73.7kB     63
4def7a01d61a        discovery-server    0.55%               161.3MiB / 512MiB   31.51%              55.7kB / 377kB      196MB / 0B          54
9619c8f2e1a0        config-server       0.30%               140.9MiB / 512MiB   27.53%
```

arey commented:
Memory limits looks like to depend of OS and configuration. I suggest to increase the memory limit to 512m. It's a good compromise.
@Xaseron you have to resolved a conflict because we have migrated to OpenZipkin.

Xaseron committed:
Merge openzipkin

Xaseron committed:
Remove obsolete links

Xaseron committed:
Fix honoring memory limits for openzipkin

Xaseron commented:
Ok, i've rebased onto master. 
In the compose-file i removed the links, because all container are by default in the same network. Links are only needed for settings aliases like:
```
    links:
      - "config-server:conf"
```

Addionally i set the environment variables for the zipkin server.

pramath01 commented:
I think we should reopen this issue if the image sizes are still in the 700–800 MB range.
 I don't remember the exact image size when we were using `eclipse-temurin:17`, but I do remember it being significantly larger (and it was consuming too much disk space inside wsl docker daemon), which is why I switched the runtime image to Alpine.
The builder stage is still based on `eclipse-temurin:17`, but the final runtime image uses `eclipse-temurin:17-jre-alpine`. The only additional change required was:

 ```dockerfile
 RUN apk add --no-cache curl
 ```
 since `curl` is needed for the health check.
 The multi-stage build now looks like this:
 * **Builder:** `eclipse-temurin:17`
 * **Runtime:** `eclipse-temurin:17-jre-alpine`
* **Additional package:** `curl` (for health checks)

 I've attached the Dockerfile diff for reference.

<img width="1518" height="830" alt="image" src="https://github.com/user-attachments/assets/b805b35f-91b0-4085-80d2-39cf6a738f8e" />

Application works as intended and build time is significantly faster. the image sizes are around 140mb. see the image below 

<img width="1025" height="167" alt="image" src="https://github.com/user-attachments/assets/9da64ee3-b8d3-4cec-ac06-b0ab7b56bd13" />
